tsdb/head_append.go

```diff
@@ -1381,17 +1387,97 @@ func (s *memSeries) appendPreprocessor(
+// histogram chunks are cut based on their size in bytes.
+// It is unsafe to call this concurrently with s.iterator(...) without holding the series lock.
+// This should be called only when appending data.
+func (s *memSeries) nativeHistogramsAppendPreprocessor(t int64, e chunkenc.Encoding, chunkDiskMapper *chunks.ChunkDiskMapper, chunkRange int64) (c *memChunk, sampleInOrder, chunkCreated bool) {
+	c = s.head()
+
+	if c == nil {
+		if len(s.mmappedChunks) > 0 && s.mmappedChunks[len(s.mmappedChunks)-1].maxTime >= t {
+			// Out of order sample. Sample timestamp is already in the mmapped chunks, so ignore it.
+			return c, false, false
+		}
+		// There is no head chunk in this series yet, create the first chunk for the sample.
+		c = s.cutNewHeadChunk(t, e, chunkDiskMapper, chunkRange)
+		chunkCreated = true
+	}
+
+	// Out of order sample.
+	if c.maxTime >= t {
+		return c, false, chunkCreated
+	}
+
+	if c.chunk.Encoding() != e {
+		// The chunk encoding expected by this append is different than the head chunk's
+		// encoding. So we cut a new chunk with the expected encoding.
+		c = s.cutNewHeadChunk(t, e, chunkDiskMapper, chunkRange)
+		chunkCreated = true
+	}
+
+	// The first sample also encodes the layout so it's expected to be larger than the rest of the samples. We consider
+	// it overhead so we remove it from the calculations below.
+	if c.chunk.NumSamples() == 1 {
+		s.nativeHistogramChunkBytesOverhead = len(c.chunk.Bytes())
+		s.nativeHistogramChunkFirstSampleTS = c.maxTime
+	}
+	targetBytes := chunkenc.TargetBytesPerNativeHistogramChunk - s.nativeHistogramChunkBytesOverhead
+	numBytes := len(c.chunk.Bytes()) - s.nativeHistogramChunkBytesOverhead
+
+	if numBytes+s.nativeHistogramChunkBytesOverhead == 0 {
+		// It could be the new chunk created after reading the chunk snapshot,
+		// hence we fix the minTime of the chunk here.
+		c.minTime = t
+		s.nextAt = rangeForTimestamp(c.minTime, chunkRange)
+	}
+
+	// If we reach 25% of a chunk's desired maximum size, predict an end time
+	// for this chunk that will try to make samples equally distributed within
+	// the remaining chunks in the current chunk range.
+	// At the latest it must happen at the timestamp set when the chunk was cut.
+	if !s.nativeHistogramChunkHasComputedEndTime && numBytes >= targetBytes/4 {
+		// s.nativeHistogramChunkFirstSampleTS can be unset if we're already past targetBytes/4 within the first 2
+		// samples. In this case, c.minTime is the start time for the s.nextAt computation instead.
+		start := s.nativeHistogramChunkFirstSampleTS
+		if start == 0 {
+			start = c.minTime
+		}
+		ratioToFull := float64(targetBytes) / float64(numBytes)
+		s.nextAt = computeChunkEndTime(start, c.maxTime, s.nextAt, ratioToFull)
+		s.nativeHistogramChunkHasComputedEndTime = true
+	}
+	// If numBytes > targetBytes*2 then our previous prediction was invalid,
+	// most likely because samples rate has changed and now they are arriving more frequently.
```

Comment:
Or the bucket count has increased?

Reply:
Ah yea, good point; there's a wider spread of valid reasons for invalid predictions--I'll make a note